### PR TITLE
Etcd client tests

### PR DESCRIPTION
**Pull request by AlexPikalov**
It contains tests for loadbalancer/etcd which are related to the issue #239.

To test successful branch of etcd Client test certificates were generated. Apart of this successful branch there are tests for following cases:

* no machines provided
* some of certificates is corrupted

There is an extra test for publisher's main loop

### CODE REVIEW

**Comment by AlexPikalov** [comment]
There is a problem with Travis CI on go 1.6 as it is for other merged PRs, so I decided to re-submit my one and to discuss it if needed

File: loadbalancer/etcd/client_test.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
+package etcd_test
+
+import (
+	etcd "github.com/coreos/etcd/client"
+	kitetcd "github.com/go-kit/kit/loadbalancer/etcd"
+	"golang.org/x/net/context"
+)
+
+import "testing"
```
Comment (peterbourgon) [inline loadbalancer/etcd/client_test.go]: Can you move this to the import block above?

File: loadbalancer/etcd/client_test.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
+package etcd_test
+
+import (
+	etcd "github.com/coreos/etcd/client"
+	kitetcd "github.com/go-kit/kit/loadbalancer/etcd"
+	"golang.org/x/net/context"
+)
```
Comment (peterbourgon) [inline loadbalancer/etcd/client_test.go]: Please group imports according to the "standard", i.e.

```go
import (
    // stdlib

    // third-party

    // go-kit/kit/*
)

File: loadbalancer/etcd/client_test.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+import "testing"
+
+func TestNoCertificateClient(t *testing.T) {
+	context := context.Background()
```
Comment (peterbourgon) [inline loadbalancer/etcd/client_test.go]: This shadows the context package. Either use `ctx`, or just pass `context.Background()` directly to the various constructors.

File: loadbalancer/etcd/client_test.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	_, err = c.GetEntries("key")
+	if err.Error() != etcd.ErrClusterUnavailable.Error() {
+		t.Fatalf("unexpected error %v", err)
+	}
+
+	ch := make(chan *etcd.Response)
+	// should fail but no idea how to check it explicitly
+	c.WatchPrefix("foo", ch)
```
Comment (peterbourgon) [inline loadbalancer/etcd/client_test.go]: What are you trying to do here? I'm sure we can figure it out :)

File: loadbalancer/etcd/client_test.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+	context := context.Background()
+	ops := &kitetcd.ClientOptions{
+		Key:    "./test_certs/host.key",
+		Cert:   "./test_certs/host.crt",
+		CaCert: "./test_certs/rootCA.crt",
+	}
+
+	_, err := kitetcd.NewClient(context, []string{}, ops)
+	if err == nil {
+		t.Fatalf("should return error if no machines provided")
```
Comment (peterbourgon) [inline loadbalancer/etcd/client_test.go]: `t.Fatal`

File: loadbalancer/etcd/client_test.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	context := context.Background()
+	ops := &kitetcd.ClientOptions{
+		Key:    "./test_certs/host.key",
+		Cert:   "boom",
+		CaCert: "./test_certs/rootCA.crt",
+	}
+
+	_, err := kitetcd.NewClient(context, []string{"http://localhost:2379"}, ops)
+	if err == nil {
+		t.Fatalf("should return an error if Certificate is corrupted")
```
Comment (peterbourgon) [inline loadbalancer/etcd/client_test.go]: `t.Fatal`

File: loadbalancer/etcd/client_test.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	context := context.Background()
+	ops := &kitetcd.ClientOptions{
+		Key:    "./test_certs/host.key",
+		Cert:   "./test_certs/host.crt",
+		CaCert: "boom",
+	}
+
+	_, err := kitetcd.NewClient(context, []string{"http://localhost:2379"}, ops)
+	if err == nil {
+		t.Fatalf("should return an error if CA Certificate is corrupted")
```
Comment (peterbourgon) [inline loadbalancer/etcd/client_test.go]: `t.Fatal`

File: loadbalancer/etcd/client_test.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	context := context.Background()
+	ops := &kitetcd.ClientOptions{
+		Key:    "./test_certs/host.key",
+		Cert:   "./test_certs/host.crt",
+		CaCert: "./test_certs/rootCA.crt",
+	}
+
+	c, err := kitetcd.NewClient(context, []string{"http://localhost:2379"}, ops)
+	if err != nil {
+		t.Fatalf("failed to create new client %v", err)
```
Comment (peterbourgon) [inline loadbalancer/etcd/client_test.go]: `t.Fatal(err)` is sufficient

File: loadbalancer/etcd/client_test.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+	}
+
+	_, err = c.GetEntries("key")
+	if err.Error() != etcd.ErrClusterUnavailable.Error() {
```
Comment (peterbourgon) [inline loadbalancer/etcd/client_test.go]: You should be able to compare values directly, i.e. 

```go
if err != etcd.ErrClusterUnavailable {
```

File: loadbalancer/etcd/client_test.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+		t.Fatalf("unexpected error %v", err)
+	}
```
Comment (peterbourgon) [inline loadbalancer/etcd/client_test.go]: I think you should be able to compare errors directly, and also use the standard `want, have` idiom, i.e.

```go
if want, have := etcd.ErrClusterUnavailable, err; want != have {
    t.Fatalf("want %v, have %v", want, have)
}
```

File: loadbalancer/etcd/publisher_test.go [hunk 11]
```diff
@@ -49,6 +50,8 @@ func TestPublisher(t *testing.T) {
 	if _, err := p.Endpoints(); err != nil {
 		t.Fatal(err)
 	}
+	// wait for the loop before stop
+	time.Sleep(time.Millisecond)
```
Comment (peterbourgon) [inline loadbalancer/etcd/publisher_test.go]: Why?

File: loadbalancer/etcd/test_certs/host.crt [hunk 12]
```diff
@@ -0,0 +1,20 @@
+-----BEGIN CERTIFICATE-----
```
Comment (peterbourgon) [inline loadbalancer/etcd/test_certs/host.crt]: Can you please include a README or something in this directory with clear, reproducible instructions on how to generate these files?

**Comment by peterbourgon** [comment]
Thanks for the contribution! Just a few things to update :)

File: loadbalancer/etcd/client_test.go (hunk 4 above)
Comment (AlexPikalov) [inline loadbalancer/etcd/client_test.go]: Well, I'm testing https://github.com/go-kit/kit/blob/master/loadbalancer/etcd/client.go#L127
In fact it fails as I expect and this branch is covered finally, but I don't like that I can not check this error was handled

File: loadbalancer/etcd/publisher_test.go (hunk 11 above)
Comment (AlexPikalov) [inline loadbalancer/etcd/publisher_test.go]: From my current understanding https://github.com/go-kit/kit/blob/master/loadbalancer/etcd/publisher_test.go#L47 stops the loop before the mock

```
func (c *fakeClient) WatchPrefix(prefix string, responseChan chan *stdetcd.Response) {
	response, ok := c.responses[prefix]
	if ok {
		responseChan <- response
	}
}
```
sends to response channel. Looks like a hack, isn't it?

File: loadbalancer/etcd/test_certs/host.crt (hunk 12 above)
Comment (AlexPikalov) [inline loadbalancer/etcd/test_certs/host.crt]: Sure. Will do.

File: loadbalancer/etcd/client_test.go (hunk 2 above)
Comment (AlexPikalov) [inline loadbalancer/etcd/client_test.go]: Alright

File: loadbalancer/etcd/client_test.go (hunk 1 above)
Comment (AlexPikalov) [inline loadbalancer/etcd/client_test.go]: Okay, I'll reformat imports according to the scheme you described

File: loadbalancer/etcd/client_test.go (hunk 10 above)
Comment (AlexPikalov) [inline loadbalancer/etcd/client_test.go]: I've tried this first. Unsuccessfully. As it goes from etcd official docs we need to use another approach of handling etcd-client errors. https://github.com/coreos/etcd/tree/master/client#error-handling

```
if cerr, ok := err.(*client.ClusterError); ok {
        // process (cerr.Errors)
}
```
So if you are good I'll use this approach in client tests.

**Comment by peterbourgon** [comment]
Ping! :) Any update here?

**Comment by peterbourgon** [comment]
I'm afraid this is abandoned :( If I got it wrong, please update and let me know!

**Pull request closed without merging**